### PR TITLE
Skeleton for kernels

### PR DESCRIPTION
**Pull request by AVHopp**
This PR adds a basic skeleton for organizing kernels.

It introduces an ABC for kernels and implements a `MaternKernel` which was so far used as default in the GPs. It also introduces a serialization test.

Note that this PR is intended to be rather minimal and just the basis for more work on kernels.

### CODE REVIEW

File: baybe/kernels/__init__.py
Comment (AdrianSosic) [inline baybe/kernels/__init__.py:1]: Would shorten + not have abbreviations in there
```suggestion
"""Kernels for Gaussian process surrogate models."""
```

File: baybe/kernels/__init__.py
Comment (AVHopp) [inline baybe/kernels/__init__.py:4]: I am not happy with the naming here, do you have other ideas of how the stuff could be called to avoid `kernels.kernels`?

File: baybe/kernels/__init__.py
Comment (AdrianSosic) [inline baybe/kernels/__init__.py:3]: The base class should not be exposed via the init file (see other packages)

File: baybe/kernels/base.py
Comment (AdrianSosic) [inline baybe/kernels/base.py:19]: why the **kwargs?

File: baybe/kernels/kernels.py
Comment (AdrianSosic) [inline baybe/kernels/kernels.py:1]: "usable" is a bit unnecessary, no?

File: baybe/kernels/kernels.py
Comment (AdrianSosic) [inline baybe/kernels/kernels.py:11]: Let's not put parameter names into docstrings (refactorability)

File: baybe/kernels/kernels.py
Comment (AdrianSosic) [inline baybe/kernels/kernels.py:17]: Avoid repeating defaults (refactorability!). The default will be automatically displayed in the doc

File: baybe/kernels/kernels.py
Comment (AdrianSosic) [inline baybe/kernels/kernels.py:16]:
```suggestion
    Only takes the values 0.5, 1.5 or 2.5. Larger values yield smoother interpolations.
```

File: tests/serialization/test_kernel_serialization.py
Comment (AdrianSosic) [inline tests/serialization/test_kernel_serialization.py:1]: Can you do the hypothesis version right away? Since you have a plain attrs class, this works pretty much out-of-the box.

**Review by AdrianSosic — CHANGES_REQUESTED** [review]
Hi @AVHopp, thanks for the draft. Here a few comments

File: baybe/kernels/__init__.py
Comment (AdrianSosic) [reply baybe/kernels/__init__.py:4]: How about "basic"?

File: baybe/kernels/base.py
Comment (AVHopp) [reply baybe/kernels/base.py:19]: This ensures that everything that you pass as keyword argument is passed to the constructor of the gpytorch kernel, no matter what exactly you pass. I thought that this was the best/most elegant way of ensuring this "passthrough", if there is a better solution I am more than happy to change this 😄

File: baybe/kernels/kernels.py
Comment (AVHopp) [reply baybe/kernels/kernels.py:1]: Just wait until I start implementing more kernels, then you'll see that this is not unnecessary 😎 

Jokes aside, will change :D

File: tests/serialization/test_kernel_serialization.py
Comment (AdrianSosic) [reply tests/serialization/test_kernel_serialization.py:1]: Yes. This is what you'll need: https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.builds

File: baybe/kernels/__init__.py
Comment (AVHopp) [reply baybe/kernels/__init__.py]: Changed

File: baybe/kernels/kernels.py
Comment (AVHopp) [reply baybe/kernels/kernels.py]: Changed

File: baybe/kernels/kernels.py
Comment (AVHopp) [reply baybe/kernels/kernels.py]: Changed.

File: baybe/kernels/kernels.py
Comment (AVHopp) [reply baybe/kernels/kernels.py]: Changed.

File: tests/serialization/test_kernel_serialization.py
Comment (AVHopp) [reply tests/serialization/test_kernel_serialization.py:1]: Why? *Our* Matern Kernel only accepts a single argument, and that is `nu`. The `**kwargs` stuff is only relevant when calling `to_gpytorch`

Comment (AVHopp) [reply tests/serialization/test_kernel_serialization.py:1]: In general, I do not see the value of hypothesis here now that I think about it. The Matern Kernel only has three allowed values, so the test would be something like the following if I understand things correctly:
```python
@st.composite
def matern_kernel(draw: st.DrawFn):
    nu = draw([0.5, 1.5, 2.5])
    return MaternKernel(nu=nu)
```

Comment (AVHopp) [reply tests/serialization/test_kernel_serialization.py:1]: And how do I even need to tell hypothesis that stuff like that should be tested? Just running `pytest` does not seem to detect the tests like `hypothesis_strategies/dataframes.py` and similar for me (at least locally)

Comment (AVHopp) [reply tests/serialization/test_kernel_serialization.py:1]: I have tried to adjust the test. I used `st.sampled_from` which seems to guarantee that it indeed samples everything from a sequence if it has less than 100 entries, so I hope that this solution is as it should be

File: baybe/kernels/__init__.py
Comment (Scienfitz) [reply baybe/kernels/__init__.py:3]: +1

File: CHANGELOG.md
Comment (Scienfitz) [inline CHANGELOG.md:9]: should also add (here or in the net PR that adds more kernels) that the kernel type can now be specified instead if being hardcoded

File: baybe/kernels/kernels.py
Comment (Scienfitz) [inline baybe/kernels/kernels.py:13]: should we add a converter that alos allows strings a la `1/2`, `3/2`, `5/2` ? At least I tend to think of these kernels parameters more in these terms

File: baybe/kernels/base.py
Comment (Scienfitz) [reply baybe/kernels/base.py:19]: could be both

additionally I wonder where the optional `nu` parameter you can specify int he new baybe `MaternKernel` class comes into play, doesnt seem to be the case?

**Review by Scienfitz — CHANGES_REQUESTED** [review]

File: baybe/kernels/base.py
Comment (AVHopp) [reply baybe/kernels/base.py:19]: True, I still need to parse `*args*` and also the fields, thanks for spotting!

Comment (AVHopp) [reply baybe/kernels/base.py]: I have attempted to fix this using the `unstructure_base` method on `self` to get all of the fields the corresponding class might have, extracting the fields, and extending `kwargs`.
Is this acceptable? It works, but I do not know if there is a safer/more elegant way.

File: CHANGELOG.md
Comment (AVHopp) [reply CHANGELOG.md:9]: Suggestion: Make it two points as follows under Added:
- `Kernel` base class allowing to specify kernel types
- `MaternKernel` can be chosen for GP surrogates

File: baybe/kernels/kernels.py
Comment (AVHopp) [reply baybe/kernels/kernels.py]: Implemented this via a converter, please resolve if the approach is suitable.

File: baybe/kernels/__init__.py
Comment (AVHopp) [reply baybe/kernels/__init__.py:3]: Will do this at the very end when rebasing and stuff like that and will keep that comment as merge-blocker. In particular, feel free to approve the PR if this is the only open conversation remaining :)

File: baybe/kernels/base.py
Comment (AVHopp) [reply baybe/kernels/base.py]: As discussed in the meeting and the chat, I have now changed this and created a corresponding utility function in `baybe.utils.basic`. Please let me know if this solution is now more favorable.

File: baybe/kernels/kernels.py
Comment (Scienfitz) [reply baybe/kernels/kernels.py]: keep in mind to add this also to hypothesis strategies

Comment (AdrianSosic) [reply baybe/kernels/kernels.py]: Please don't add to the strategy itself but rather add a test to the `alternative_creation` folder (that's the purpose of it). I followed the principle that the strategy should merely span the space of possibilities and are minimal in that sense. I think this keeps the code easier to maintain

Comment (AVHopp) [reply baybe/kernels/kernels.py]: Done. Added two tests for the string based representation of the corresponding fractions and floats and one strategy for creating `Fraction` objects. Please resolve if this is what you had in mind :)

File: CHANGELOG.md
Comment (AVHopp) [reply CHANGELOG.md:9]: Changed it accordingly, please either resolve or propose different formulation.

File: tests/hypothesis_strategies/alternative_creation/test_kernels.py
Comment (Scienfitz) [inline tests/hypothesis_strategies/alternative_creation/test_kernels.py:26]: whats the point of this test and why is it in alternative construction?

File: baybe/kernels/base.py
Comment (Scienfitz) [inline baybe/kernels/base.py:26]: why not "args" too ? (since these are the function defaults anyway why do you set them here explicitly?)

File: baybe/utils/basic.py
Comment (Scienfitz) [inline baybe/utils/basic.py:95]: why not "args" too?

File: tests/hypothesis_strategies/alternative_creation/test_kernels.py
Comment (Scienfitz) [inline tests/hypothesis_strategies/alternative_creation/test_kernels.py:21]: as this will shrink towards the first entry I would put the 5/2 first as its our default

File: baybe/kernels/kernels.py
Comment (Scienfitz) [inline baybe/kernels/kernels.py:18]: not sure why you chose to detailed docstring this valdiator, but to save space at least the last sentence `Note that...` can go as its trivial

**Review by Scienfitz — APPROVED** [review]
this was the perfect atomic PR size 👍

File: CHANGELOG.md
Comment (AdrianSosic) [inline CHANGELOG.md:10]: you can add:
`hypothesis` strategy and roundtrip test for kernels

File: baybe/kernels/kernels.py
Comment (AdrianSosic) [reply baybe/kernels/kernels.py:18]: Agree with Martin but would go even a step further. Will add explanation in my review

File: tests/hypothesis_strategies/alternative_creation/test_kernels.py
Comment (AVHopp) [reply tests/hypothesis_strategies/alternative_creation/test_kernels.py:26]: In theory, users can also provide other fractions like 6/4 which are mathematically valid choices. I thus wanted to test whether the kernel is robust against arbitrary fractions which are de facto just 1/2, 3/2, 5/2. If this is unnecessary, we can also just remove it.

File: baybe/kernels/base.py
Comment (AVHopp) [reply baybe/kernels/base.py:26]: I'm setting them explicitly since I always prefer to have stuff explicitly. Args did not show up in fields_dict, hence no reason to remove it. Could however also add this (also to the default arguments if you prefer)

File: baybe/kernels/kernels.py
Comment (AVHopp) [reply baybe/kernels/kernels.py:18]: I have to be honest: The reason is because I am personally not happy with the way we comment on validators, default, converters and so on. I would appreciate full docstrings for those, even if they are not necessary. Might be a personal issue resp. due to me not having the same knowledge about all attrs/cattrs stuff like you. This is also the reason why I did not push this (in addition to this being much more work)
But fully agree: The "Note that...." part can go 👍  Will remove once  @AdrianSosic has finished his work here (or feel free to already remove yourself)

File: baybe/utils/basic.py
Comment (AVHopp) [reply baybe/utils/basic.py:95]: See comment at the other thread. Happy to adjust, although I am not sure if it would have any influence.

File: tests/hypothesis_strategies/alternative_creation/test_kernels.py
Comment (AVHopp) [reply tests/hypothesis_strategies/alternative_creation/test_kernels.py:21]: Will change if we decide to keep this.

File: tests/serialization/test_kernel_serialization.py
Comment (AdrianSosic) [inline tests/serialization/test_kernel_serialization.py:1]:
```suggestion
"""Test serialization of kernels."""
```

File: tests/serialization/test_kernel_serialization.py
Comment (AdrianSosic) [inline tests/serialization/test_kernel_serialization.py:13]:
```suggestion
    assert kernel == kernel2, (kernel, kernel2)
```

File: tests/hypothesis_strategies/kernels.py
Comment (AdrianSosic) [inline tests/hypothesis_strategies/kernels.py:7]: We adopted the hypothesis naming convention for strategies, which is "snake case plural".
```suggestion
matern_kernels = st.builds(MaternKernel, st.sampled_from((0.5, 1.5, 2.5)))
```
Of course, you need to change also downstream

File: CHANGELOG.md
Comment (AdrianSosic) [inline CHANGELOG.md:1]: General: in your docstrings you often write "str". If neither of you have a preference, I would avoid abbreviations as a general rule

File: tests/hypothesis_strategies/alternative_creation/test_kernels.py
Comment (AdrianSosic) [reply tests/hypothesis_strategies/alternative_creation/test_kernels.py:26]: I also don't get the point to be honest. Seems like "mit Kanonen auf Spatzen geschossen" to me. Having the test is fine I think, but using a normal parametrization containing the three fraction fully suffices. (Remember: your test now runs like 100 times or so with repeated input)

**Review by AdrianSosic — CHANGES_REQUESTED** [review]
Here you go 🙃

File: tests/serialization/test_kernel_serialization.py
Comment (AdrianSosic) [reply tests/serialization/test_kernel_serialization.py:1]: I think this can be resolved since we have another active thread

File: baybe/kernels/kernels.py
Comment (Scienfitz) [reply baybe/kernels/kernels.py:36]:
```suggestion
                f"Provided value {nu} of 'nu' could not be interpreted as fraction"
```

File: tests/hypothesis_strategies/alternative_creation/test_kernels.py
Comment (AVHopp) [reply tests/hypothesis_strategies/alternative_creation/test_kernels.py:26]: Ok, I will change this to a test similar in style to the others and just test for the three `Fraction` objects.

File: baybe/kernels/kernels.py
Comment (AVHopp) [reply baybe/kernels/kernels.py]: Changed it as discussed in chat.

File: baybe/kernels/kernels.py
Comment (AVHopp) [reply baybe/kernels/kernels.py]: Done.

File: baybe/kernels/kernels.py
Comment (AVHopp) [reply baybe/kernels/kernels.py]: This was changed as discussed in our chat, so I'll resolve here. In case there is anything to discuss regarding the conversion function in general, feel free to resolve one of the threads discussing it.

File: tests/hypothesis_strategies/alternative_creation/test_kernels.py
Comment (AVHopp) [reply tests/hypothesis_strategies/alternative_creation/test_kernels.py:26]: Changed. Since I am not sure whether this is no as you would prefer resp. if we should keep the test in general, I will not resolve now until you are happy. Will however resolve other threads related to these tests such that we have a single point of discussion.

File: tests/hypothesis_strategies/alternative_creation/test_kernels.py
Comment (AVHopp) [reply tests/hypothesis_strategies/alternative_creation/test_kernels.py:21]: Resolved as there is another thread to discuss this test.

File: baybe/utils/basic.py
Comment (AVHopp) [reply baybe/utils/basic.py:104]: Fully agree, this is inconsistent. Isn't `attributes` correct? I literally call `getattr` 😄 Do you have any other idea/suggestion?

File: baybe/utils/basic.py
Comment (AVHopp) [reply baybe/utils/basic.py]: Good idea(s). Will change :)

File: tests/serialization/test_kernel_serialization.py
Comment (AVHopp) [reply tests/serialization/test_kernel_serialization.py]: Changed.

File: CHANGELOG.md
Comment (AVHopp) [reply CHANGELOG.md]: Since there is no comment/complaint regarding this, I interpret this as fine.

File: baybe/kernels/base.py
Comment (AVHopp) [reply baybe/kernels/base.py]: Since there was no comment regarding this issue in both of your PRs from yesterday, I'll mark this as resolved. There is also another open comment regarding this function, so in case there is something to discuss, we can do it there.

File: baybe/utils/basic.py
Comment (AVHopp) [reply baybe/utils/basic.py]: Changed as you proposed, will hence resolve.

File: CHANGELOG.md
Comment (AVHopp) [reply CHANGELOG.md:1]: I had a look and did not spot any further occurrences of "str" in my docstrings. Will hence resolve.

File: baybe/utils/basic.py
Comment (AVHopp) [reply baybe/utils/basic.py]: I have changed this to `get_init_arguments`, since this is what the function actually does - it gets the arguments of the `__init__` function. I tried to make the role of `filter` clear in the docstring. Please resolve if you are happy with this.
Also, feel free to use this thread for other comments on this function (resolved several that were outdated resp. not being commented on during your last reviews)

File: tests/hypothesis_strategies/alternative_creation/test_kernels.py
Comment (AdrianSosic) [reply tests/hypothesis_strategies/alternative_creation/test_kernels.py:11]: I think this is fine now! Will resolve

File: CHANGELOG.md
Comment (AVHopp) [inline CHANGELOG.md:12]: @AdrianSosic @Scienfitz all changes included :) Two things are still kind of open, I have kept two threads discussing those open since there were intersections/comments out of date and so on:
1. The tests: I have removed the strategy for creating all of the `Fraction` objects and instead just included a test with the three correponding fraction objects.
2. The function for getting the stuff from init: I have renamed it to `get_init_arguments` since this is what it actually does and adjusted variable names and docstring. Since this will also be used in other contexts, I want to make sure that the naming I chose is correct and makes sense.

Also, the _convert_nu function underwent some major change, but since this was discussed in our internal chat, I did not keep a thread on this - just fyi. Once you have giving your "gucci" regarding the content, I will clean the commit history, force-push and merge.

File: baybe/utils/basic.py
Comment (AdrianSosic) [reply baybe/utils/basic.py]: Hi @AVHopp, first of all, thanks for the changes. I hate to be the annoying guy, but unfortunately there are still quite many issues with your function:
* The optional filter does not really work, i.e. it's not really optional. Try to run this code and you'll see what I mean:
   ```python
  import gpytorch
  
  from baybe.kernels import MaternKernel
  from baybe.utils.basic import get_init_arguments
  
  
  x = get_init_arguments(
      target=gpytorch.kernels.MaternKernel, ignored_args=("self", "kwargs")
  )
  print(x)
  
  y = get_init_arguments(
      target=gpytorch.kernels.MaternKernel,
      ignored_args=("self", "kwargs"),
      filter=MaternKernel(),
  )
  print(y)
  ```
* This one is on me: I forgot that `filter` is actually a python keyword, i.e. we shouldn't use it (surprised that no linter complained). Any alternative ideas? Maybe `reference`?
* Would be good if you could add a fully defined output type (i.e. specify the dict keys and values in the type hint)
* The way you typed/documented your function and the way you use it isn't really aligned. You speak of classes as inputs, but you actually provide an object as input to the filter argument. 
* I can live with `attributes` (though technically it should be parameters), but in the return you again talk about parameters -> still inconsistent. Either change it there or rename the function to `get_init_params` and use "parameter" everywhere
* Also, if you just read the documentation (and the function name), it is absolutely unclear what the output dictionary even contains or why it is a dictionary in the first place. I think you need to specify that in the output docstring

File: baybe/kernels/basic.py
Comment (AdrianSosic) [inline baybe/kernels/basic.py:19]:
```suggestion
        The float representation of the given input.
```
(you never mentioned "parameter" above)

File: baybe/kernels/basic.py
Comment (AdrianSosic) [inline baybe/kernels/basic.py:23]:
```suggestion
        ValueError: If the input was provided as string but could not be interpreted as
            fraction.
```
Good that you added the `/` to the signature, but the whole point of this is that the name of the parameter is hidden from the user. So it makes no sense to reference it in the doc/error message

File: baybe/kernels/basic.py
Comment (AdrianSosic) [inline baybe/kernels/basic.py:31]:
```suggestion
                f"The provided input '{value}' could not be interpreted as a fraction."
```
(reason: see comment above)

**Review by AdrianSosic — CHANGES_REQUESTED** [review]

File: baybe/utils/basic.py
Comment (AdrianSosic) [reply baybe/utils/basic.py]: Ah, I just also see that mypy failed in your pipeline run, which confirms the typing issue 👍🏼

Comment (AdrianSosic) [reply baybe/utils/basic.py]: (I guess that make me somewhat less the bad guy since we can now just blame mypy 😄 )

Comment (AVHopp) [reply baybe/utils/basic.py]: Proposed solution (not yet pushed): Would this work?
```python
def get_attributes_based_on_init_parameters(
    target_class: type,
    object: Any,
    ignored_attrs: tuple[str, ...] = ("self", "kwargs", "args"),
) -> dict[str, Any]:
    """Get attributes of an object based on the parameters of a class` ``__init__``.

    Args:
        target_class: The target class.
        object: The object whose attributes should be returned.
        ignored_attrs: Arguments that will be ignored.

    Returns:
        The dictionary of parameters.
    """
    # Get the parameters of the target signature
    params = signature(target_class.__init__).parameters
    return {p: getattr(object, p) for p in params if p not in ignored_attrs}
    ```

Comment (AVHopp) [reply baybe/utils/basic.py]: Good point, will make it optional.
@AdrianSosic Could you very quickly comment on that piece of code? It seems like your confirmation (since you raised that issue) is the only remaining piece blocking this PR

Comment (AdrianSosic) [reply baybe/utils/basic.py]: Ok, let me quickly test this. In the meantime: I don't understand @Scienfitz comment!? In the posted snippet, the parameter IS already optional (i.e. has a default, to be precise). So what exactly are we talking about?

Comment (AVHopp) [reply baybe/utils/basic.py]: I guess it referes to the type hint which is still missing.

Comment (AdrianSosic) [reply baybe/utils/basic.py]: The type hint is fine IMO, but I'll post a refined version in a minute

Comment (AVHopp) [reply baybe/utils/basic.py]: That looks good to me. Will add that, see if everything works as expected and then resolve this thread. Thanks to bot of you for the feedback!

Comment (AVHopp) [reply baybe/utils/basic.py]: Just one question: Why the underscore at the end of `callable_`?

Comment (AdrianSosic) [reply baybe/utils/basic.py]: Some problems as with `filter`. `callable` is a python keyword, unfortunately. Feel free to replace this with any other appropriate name if you can find one 👍🏼

File: CHANGELOG.md
Comment (AVHopp) [reply CHANGELOG.md:13]: @Scienfitz @AdrianSosic final rebase and git history cleaning done, the PR is now also down to a reasonable number of commits :D Please given your final approve/comments and feel free to ping me for merge or merge yourselves

**Review by AdrianSosic — APPROVED** [review]
Very nice, thanks a lot for fixing all the nitty-gritty details and even going the extra mile of refactoring, @AVHopp. Leave the honor of merging to you :D

File: CHANGELOG.md
Comment (AdrianSosic) [reply CHANGELOG.md:13]: Gorgeous 🐕 🥇 doge much approve. Ready for meeeerge 🚀

**Pull request merged**
